src/shared/api/server/block/index.ts

```diff
@@ -0,0 +1,103 @@
+import { NextRequest, NextResponse } from 'next/server';
+import { pindexer } from '@/shared/database';
+import {
+  BlockSummaryApiResponse,
+  BatchSwapSummaryDisplay,
+} from '@/shared/api/server/block/types.ts';
+import { AssetId } from '@penumbra-zone/protobuf/penumbra/core/asset/v1/asset_pb';
+import { ChainRegistryClient, Registry } from '@penumbra-labs/registry';
+import { getDisplayDenomExponent } from '@penumbra-zone/getters/metadata';
+import { pnum } from '@penumbra-zone/types/pnum';
+import { BatchSwapSummary } from '@/shared/database/schema';
+import { hexToUint8Array, base64ToHex } from '@penumbra-zone/types/hex';
+import { joinLoHi, LoHi } from '@penumbra-zone/types/lo-hi';
+
+function isLoHi(value: unknown): value is LoHi {
+  return typeof value === 'object' && value !== null && 'lo' in value;
+}
+
+interface AssetWithInner {
+  inner: string;
+}
+
+function isAsset(value: unknown): value is AssetWithInner {
+  return (
+    typeof value === 'object' &&
+    value !== null &&
+    'inner' in value &&
+    value.inner !== null &&
+    typeof (value as AssetWithInner).inner === 'string'
+  );
+}
+
+export const getBatchSwapDisplayData =
+  (registry: Registry) =>
+  (batchSwapSummary: BatchSwapSummary): BatchSwapSummaryDisplay => {
+    if (!isLoHi(batchSwapSummary.input) || !isLoHi(batchSwapSummary.output)) {
+      throw new Error('Invalid input or output: expected LoHi type');
+    }
+
+    if (!isAsset(batchSwapSummary.asset_start) || !isAsset(batchSwapSummary.asset_end)) {
+      throw new Error('Invalid asset_start or asset_end');
+    }
+
+    const startAssetId = new AssetId({
+      inner: hexToUint8Array(base64ToHex(batchSwapSummary.asset_start.inner)),
+    });
+    const startMetadata = registry.getMetadata(startAssetId);
+    const startExponent = getDisplayDenomExponent.optional(startMetadata) ?? 0;
+
+    const endAssetId = new AssetId({
+      inner: hexToUint8Array(base64ToHex(batchSwapSummary.asset_end.inner)),
+    });
+    const endMetadata = registry.getMetadata(endAssetId);
+    const endExponent = getDisplayDenomExponent.optional(endMetadata) ?? 0;
+
+    const inputBigInt = joinLoHi(batchSwapSummary.input.lo, batchSwapSummary.input.hi);
+    const outputBigInt = joinLoHi(batchSwapSummary.output.lo, batchSwapSummary.output.hi);
+
+    return {
+      startAsset: startMetadata.toJson(),
+      endAsset: endMetadata.toJson(),
```

Comment:
question: doesn't `serialize()` function apply here? Wandering why is it serialized by-hand here?

Reply:
Yes! Tbh, I wasn't aware of this piece of code, but I have replaced it and learned a few things along the way.